pkg/kubectl/scale.go

```diff
@@ -516,3 +519,93 @@ func (scaler *DeploymentScaler) Scale(namespace, name string, newSize uint, prec
+	}
+	return nil
+}
+
+// GenericScaler can update scales for resources in a particular namespace
+// TODO(o0lyn0mial): when the work on GenericScaler is done, don't
+// export the GenericScaler. Instead use ScalerFor method for getting the Scaler
+// also update the UTs
+type GenericScaler struct {
+	scaleNamespacer scaleclient.ScalesGetter
+	targetGR        schema.GroupResource
+}
+
+var _ Scaler = &GenericScaler{}
+
+// ScaleSimple updates a scale of a given resource. It returns the resourceVersion of the scale if the update was successful.
+func (s *GenericScaler) ScaleSimple(namespace, name string, preconditions *ScalePrecondition, newSize uint) (updatedResourceVersion string, err error) {
+	scale, err := s.scaleNamespacer.Scales(namespace).Get(s.targetGR, name)
+	if err != nil {
+		return "", ScaleError{ScaleGetFailure, "", err}
+	}
+	if preconditions != nil {
+		if err := preconditions.validateGeneric(scale); err != nil {
+			return "", err
+		}
+	}
+
+	scale.Spec.Replicas = int32(newSize)
+	updatedScale, err := s.scaleNamespacer.Scales(namespace).Update(s.targetGR, scale)
+	if err != nil {
+		if errors.IsConflict(err) {
+			return "", ScaleError{ScaleUpdateConflictFailure, scale.ResourceVersion, err}
+		}
+		return "", ScaleError{ScaleUpdateFailure, scale.ResourceVersion, err}
+	}
+	return updatedScale.ResourceVersion, nil
+}
+
+// Scale updates a scale of a given resource to a new size, with optional precondition check (if preconditions is not nil),
+// optional retries (if retry is not nil), and then optionally waits for the status to reach desired count.
+func (s *GenericScaler) Scale(namespace, resourceName string, newSize uint, preconditions *ScalePrecondition, retry, waitForReplicas *RetryParams) error {
+	if preconditions == nil {
+		preconditions = &ScalePrecondition{-1, ""}
+	}
+	if retry == nil {
+		// make it try only once, immediately
+		retry = &RetryParams{Interval: time.Millisecond, Timeout: time.Millisecond}
+	}
+	cond := ScaleCondition(s, preconditions, namespace, resourceName, newSize, nil)
+	if err := wait.Poll(retry.Interval, retry.Timeout, cond); err != nil {
+		return err
+	}
+	if waitForReplicas != nil {
+		actualScale, err := s.scaleNamespacer.Scales(namespace).Get(s.targetGR, resourceName)
+		if err != nil {
+			return err
+		}
+		err = wait.Poll(
+			waitForReplicas.Interval,
+			waitForReplicas.Timeout,
+			scaleHasDesiredReplicas(s.scaleNamespacer, s.targetGR, resourceName, namespace, actualScale))
+		if err == wait.ErrWaitTimeout {
+			return fmt.Errorf("timed out waiting for %q to be synced", resourceName)
+		}
+		return err
+	}
+	return nil
+}
+
+// scaleHasDesiredReplicas returns a condition that will be true if and only if the desired replica
+// count for a scale (Spec) equals its updated replicas count (Status)
+func scaleHasDesiredReplicas(sClient scaleclient.ScalesGetter, gr schema.GroupResource, resourceName string, namespace string, expectedScale *autoscalingapi.Scale) wait.ConditionFunc {
```

Comment:
not sure if passing `expectedScale` makes sense.

Reply:
> not sure if passing expectedScale makes sense.

Seems like you just want `desiredReplicas` right?

Reply:
yeah, we could pass `desiredReplicas` as a number (uint) and then make the following condition in `scaleHasDesiredReplicas` method:

`return actualScale.Spec.Replicas == actualScale.Status.Replicas &&
			desiredReplicas == actualScale.Status.Replicas, nil`

does it make sense ?